e2e test for node out of disk conditions

madhusudancs opened the pull request:
This is an e2e test for PRs #16178 and #16179 that implements changes proposed in issue #4135.

madhusudancs committed:
Implement e2e test for node out of disk condition.

The ability to SSH into individual nodes, described by api.Node, and running
commands on them appears to be useful for other e2e tests too. So, move
issueSSHCommand utility function to e2e util file.

Also, modify waitForNodeToBe e2e util function to take any node condition. The
current implementation only takes api.NodeReady condition into consideration.
This change makes the function take any node condition.

madhusudancs commented:
@davidopp Please review.

I made this is a separate PR (for e2e tests since the pkg/ changes are already reviewed). Only e2e bits need review.

davidopp commented on test/e2e/util.go:
```diff
@@ -1918,12 +1937,28 @@ func isNodeReadySetAsExpected(node *api.Node, wantReady bool) bool {
 // ready or unknown).
-func waitForNodeToBe(c *client.Client, name string, wantReady bool, timeout time.Duration) bool {
-	Logf("Waiting up to %v for node %s readiness to be %t", timeout, name, wantReady)
+func waitForNodeToBe(c *client.Client, name string, condition api.NodeConditionType, wantTrue bool, timeout time.Duration) bool {
```
update the function's comment to reflect the new behavior

madhusudancs replied:
Done.

davidopp commented on test/e2e/util.go:
```diff
@@ -1918,12 +1937,28 @@ func isNodeReadySetAsExpected(node *api.Node, wantReady bool) bool {
 	return false
 }
 
+func isNodeOutOfDiskSetAsExpected(node *api.Node, wantOutOfDisk bool) bool {
```
I'm not one to usually insist on refactoring common code, but in this case I think it's warranted (this function and isNodeReadySetAsExpected() are identical except for which condition they check) -- just rename isNodeReadySetAsExpected() to isNodeConditionSetAsExpected() and have it take a cond.Type as an additional argument. Then you don't need to add a funciton.

madhusudancs replied:
Done.

Sorry about that. I was afraid to refactor more code, i.e. code that depended on existing isNodeReadySetAsExpected().

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+// 1. Fill disk space on all nodes except one. One node is left out so that we can schedule pods
+//    on that node. Arbitrarily choose that node to be node with index 0.
+// 2. Get the CPU capacity on unfilled node.
+// 3. Divide it into number of pods we want to schedule - 1. We want to schedule 3 pods, so
```
I didn't understand what the "1." and "2." mean.

davidopp replied:
Oh, now I understand. I would suggest
s/number of pods we want to schedule - 1/one less than the number of pods we want to schedule./

madhusudancs replied:
Done.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+// Reason: although there is CPU available on other nodes, pods are not schedulable on those nodes
+//         because the other nodes are out of disk.
+//
+// TODO(madhusudancs): Today, we use CPU instead of directly using storage as resource constraint
```
Once we have local disk storage as a schedulable resource, we actually won't need this feature at all, because the scheduler will prevent scheduling pods onto nodes where free disk space < requested disk space. (We'll still want the OOD nodeStatus, for debugging purposes, but there won't be any need for it to feed into the scheduler anymore.) So this TODO is not something we would do.

madhusudancs replied:
Oh Ok. Makes sense. 

I guess when we make storage a schedulable resource, we will remove this feature and hence the e2e test. So I just removed the TODO entirely.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+		for i := 0; i < numNodeConditionsPods; i++ {
+			id := podId{
+				name:  fmt.Sprintf("pod-nodecondition-%d", i),
+				value: strconv.Itoa(time.Now().Nanosecond()),
```
erictune ran into a problem using Nanosecond() in a tight loop in a different test (it returned the same value on two different loop iterations), so instead you should use a random number generator to get this value

davidopp replied:
Actually, do you even need this? The label you create in createPod() isn't used anywhere. I would just remove it.

madhusudancs replied:
I was using it in the loop below to check whether a pod was assigned to a node. But I can do that without labels, by just calling podClient.Get() instead of podClient.List().

Removed labels entirely. And the podId struct that came with it. Simplifies the code a little bit.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+		expectNoError(err)
+
+		Expect(schedEvents.Items).ToNot(BeEmpty())
+	})
```
Could you add one more step to the test, where you recover the disk space on one of the full machines and verify that the pending pod schedules there?

madhusudancs replied:
Done.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+// 3. Divide it into number of pods we want to schedule - 1. We want to schedule 3 pods, so
+//    divide CPU capacity by 2.
+// 4. Request the divided capacity for each pod.
+// 5. Attempt to schedule the final pod and observe that it fails.
```
I would replace this with:
Observe that 2 of the pods schedule onto the machine whose disk is not full, and the remaining pod stays pending and does not schedule onto the machines whose disks are full (nor the machine with the other two pods, since there is not enough free CPU capacity there).

madhusudancs replied:
Done. With a bit of alteration.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:  "nginx",
```
nginx is probably not a great choice of names, considering that the image is "pause"
maybe
Name: "pause",
would be better

madhusudancs replied:
Fixed. 

Sorry, this is a left over from the previous code where I was using an nginx image.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+						Requests: api.ResourceList{
+							// Request enough CPU to fit only two pods on a given node.
+							api.ResourceCPU:     *resource.NewMilliQuantity(milliCPU, resource.DecimalSI),
+							api.ResourceStorage: *resource.NewQuantity(50*gb, resource.DecimalSI),
```
Is there a reason you have this? We don't do anything with storage resource currently AFAIK. I would just delete it to avoid confusing someone who might read this test.

madhusudancs replied:
Done.

This is a leftover from my misunderstanding of storage as a schedulable resource.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+func availSize(c *client.Client, node api.Node) (uint64, error) {
```
add a function-level comment

madhusudancs replied:
Done.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+func fillDiskSpace(c *client.Client) {
+	// Get all nodes, and kick off the test on each.
```
I think you just want to say "fill up the disk on every node except the one with index 0."
Also I think we've been putting function-level comments above the function.

madhusudancs replied:
Fixed.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+func recoverDiskSpace(c *client.Client) {
+	// Get all nodes, and kick off the test on each.
```
Fix comment and move above function

madhusudancs replied:
Done.

davidopp commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+	expectNoError(err)
+	Expect(len(nodelist.Items)).ToNot(BeZero())
+
+	// Fill disk space in all the nodes except the first node (index 0). One node has to
```
Delete this comment

madhusudancs replied:
Done.

ncdc commented:
@kubernetes/rh-cluster-infra

pmorie commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+	})
+})
+
+func createPod(c *client.Client, ns string, id podId, milliCPU int64) {
```
nit: `createPod` is pretty general for a function that is going to be in the internal namespace for the package -- `createOutOfDiskPod` maybe?

madhusudancs replied:
Done.

pmorie commented on test/e2e/nodeconditions.go:
```diff
@@ -0,0 +1,249 @@
+			createPod(c, ns, id, podCPU)
+			podIds = append(podIds, id)
+		}
+		defer deletePods(c, ns)
```
`Framework.afterEach` deletes the namespace.  You should be able to remove this.

madhusudancs replied:
Removed this and the function definition.

madhusudancs commented:
@davidopp @pmorie I think I have addressed all the comments. PTAL.

Also, I changed the e2e test file name from nodeconditions.go to nodeoutofdisk.go. Every time I read this code, my hope of this file becoming a generic e2e test file for all the nodeconditions is becoming lesser. So I think calling it nodeoutofdisk.go is better.

davidopp commented on test/e2e/nodeoutofdisk.go:
```diff
@@ -0,0 +1,246 @@
+//    choose that node to be node with index 1.
+// 7. Observe that the pod in pending status schedules on that node.
+//
+// Reason: Initially when the disk space is filled on all but one node, the last pod fails to
```
I think this paragraph is redundant and can be deleted.

madhusudancs replied:
Done.

davidopp commented on test/e2e/nodeoutofdisk.go:
```diff
@@ -0,0 +1,246 @@
+		}
+
+		By("Waiting for all but one pods to be scheduled")
+		// Wait for all pods except the last one to be running
```
Please add
TODO: This is a bit fragile because there's no guarantee the scheduler will try to schedule the pods in order (thus there's no guarantee the last pod is the one that will stay pending).

I wouldn't bother fixing it right now.

madhusudancs replied:
Done.

davidopp commented on test/e2e/util.go:
```diff
@@ -1891,38 +1910,37 @@ func checkPodsRunningReady(c *client.Client, ns string, podNames []string, timeo
-// ready or unknown).
-func waitForNodeToBe(c *client.Client, name string, wantReady bool, timeout time.Duration) bool {
-	Logf("Waiting up to %v for node %s readiness to be %t", timeout, name, wantReady)
+// waitForNodeToBe returns whether node name's condition state matches wantTrue
```
put "name's" in quotes to make this a little easier to parse (I realize you didn't change that part of the comment)

madhusudancs replied:
Added quotes.

You are right. It is still a little difficult to parse even with this change, it is at least better than before. I thought about making it better, but none of the alternatives I could think of were better than this comment.

davidopp commented:
Looks good, just a couple of minor comments.

madhusudancs commented:
@davidopp Addressed the comments and rebased to master (other PRs were merged to master). PTAL.

davidopp commented:
LGTM

davidopp commented:
Oh, actually, before I add the LGTM label, please squash the commits.

madhusudancs commented:
Squashed the commits.

davidopp commented:
LGTM

davidopp commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot unit test this please

a-robinson commented:
@k8s-bot test this please

davidopp commented:
@k8s-bot test this please

a-robinson commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@madhusudancs It seems that the test you're introducing failed on Jenkins:
"Kubernetes e2e suite.NodeOutOfDisk runs out of disk space"

I'm temporarily removing lgtm - can you please take a look why this hapened?

madhusudancs commented:
@wojtek-t suite.NodeOutOfDisk has failed in some runs, but other tests are failing too. It's still not clear why. I am investigating.

madhusudancs commented:
@davidopp Please take a look at the last commit that I pushed today. If you LGTM, I will fold it into its previous commit.

davidopp commented on test/e2e/nodeoutofdisk.go:
```diff
@@ -0,0 +1,245 @@
+
+		nodelist, err := listNodes(c, labels.Everything(), fields.Everything())
+		expectNoError(err, "Error retrieving nodes")
+		Expect(len(nodelist.Items)).To(BeNumerically(">", 2))
```
In BeforeEach you checked `>1` rather than `>2` -- I think `>1` is correct since the test can work with just two nodes? Either way, these should be the same number (maybe make it a constant).

madhusudancs replied:
Yeah, you are right. It should have been >= 1 and > 1 in the first place. My intention was to make sure that there was at least one node in BeforeEach because one node had to be left out from filling and then make sure that there were more than one node here. But it doesn't make much sense to have >=1 in BeforeEach when we cannot have just one node here. Fixing it.

davidopp commented:
Looks great! I just had one nit. Once you fix that, you can squash the commits.

madhusudancs commented:
@k8s-bot test this please

madhusudancs commented:
@k8s-bot test this please.

(Running it again, just to make sure).

davidopp commented:
@k8s-bot test this please

davidopp commented:
LGTM

madhusudancs commented:
@k8s-bot test this please

madhusudancs commented:
@k8s-bot test this please

(Connecting to docker daemon is failing, so trying again).

madhusudancs commented:
I talked to @mikedanese about these e2e test failures and he explained how PR Jenkins runs these tests in parallel. He also showed me how to disable my test when e2e suite is run in parallel.

I am disabling this test for parallel runs by adding it to GCE_PARALLEL_SKIP_TESTS in hack/jenkins/e2e.sh. Quoting @mikedanese, "these tests are still going to fail in the PR because PR jenkins looks for hack/jenkins/e2e.sh in the master branch."

Just ran the entire test suite locally on a new cluster using this command:
```
go run ./hack/e2e.go -v --test '--test_args=--ginkgo.skip=Autoscaling\sSuite|Skipped|Restart\sshould\srestart\sall\snodes|Example|Reboot|ServiceLoadBalancer|DaemonRestart\sController\sManager|Daemon\sset\sshould\srun\sand\sstop\scomplex\sdaemon|Resource\susage\sof\ssystem\scontainers|allows\sscheduling\sof\spods\son\sa\sminion\safter\sit\srejoins\sthe\scluster|deployment.*\sin\sthe\sright\sorder|SchedulerPredicates\svalidates\sMaxPods\slimit|Nodes\sResize|resource\susage\stracking|monotonically\sincreasing\srestart\scount|Garbage\scollector\sshould|KubeProxy\sshould\stest\skube-proxy|cap\sback-off\sat\sMaxContainerBackOff'
```

All tests passed.


• [SLOW TEST:50.667 seconds]
Kubectl client
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:911
  Guestbook application
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:143
    should create and stop a working application [Conformance]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:142
------------------------------

Ran 126 of 197 Specs in 4612.962 seconds
SUCCESS! -- 126 Passed | 0 Failed | 2 Pending | 69 Skipped PASS

Ginkgo ran 1 suite in 1h16m53.251964796s
Test Suite Passed
2015/11/05 05:54:10 e2e.go:305: Step 'Ginkgo tests' finished in 1h16m54.305508287s

PTAL.

mikedanese commented:
Yup. An unfortunate consequence of our setup. Makes it hard to add e2e tests that are not safe for parallel execution. cc @ixdy

davidopp commented:
LGTM

davidopp commented:
@k8s-bot test this please

madhusudancs commented:
@davidopp I talked to @ixdy about running this e2e in Jenkins in non-parallel mode. He says there isn't really a way to run this before submission, but once the PR is merged to master the post-submits should automatically pick this up and run the tests in non-parallel Jenkins.

piosz commented:
@k8s-bot e2e test this please

mikedanese commented:
@ixdy should we just manually merge this? Seems like its not currently possible to add parallel flaky e2e tests and have mergebot merge them.

ixdy commented:
@mikedanese well, mergebot will happily merge (as long as things all pass), though it obviously won't test the new tests.

@k8s-bot e2e test this please, now that your quota is fixed

wojtek-t commented:
@madhusudancs - can you please rebase?

madhusudancs commented:
@wojtek-t Rebased.

wojtek-t commented:
@madhusudancs - thanks!

ashcrow commented:
e2e flake trying to execute up.

```
2015/11/17 18:48:49 e2e.go:309: Error running up: exit status 2
2015/11/17 18:48:49 e2e.go:305: Step 'up' finished in 12m7.853636211s
2015/11/17 18:48:49 e2e.go:141: Error starting e2e cluster. Aborting.
```

mikedanese commented:
@k8s-bot ok to test

madhusudancs commented:
@ixdy e2e tests are still failing. Is this because the e2e test is added to GCE_PARALLEL_SKIP_TESTS in this commit and PR jenkins looks at the master for this config?

mikedanese commented:
@ixdy it is running the new test which is the problem since the test is disruptive. It's not picking up that it shouldn't run the test. AFAIK, there's no way add disruptive tests in a way such the ci will be green/mergebot will merge.

mikedanese commented:
Should @madhusudancs split the hack/jenkins/e2e.sh change into a separate PR that we can merge before this goes in to get ci green?

ixdy commented on hack/jenkins/e2e.sh:
```diff
@@ -197,6 +197,7 @@ GCE_PARALLEL_SKIP_TESTS=(
     "Resource\susage\sof\ssystem\scontainers"
     "SchedulerPredicates"
     "resource\susage\stracking"
+    "NodeOutOfDisk",
```
The problem is this comma. There should not be a comma here. If you look at the skiplist arguments to ginkgo, it ends up being
```
...|NodeOutOfDisk,|...
```
which won't match the new test suite.

madhusudancs replied:
Oops, not entirely sure where I got that comma from, can't even say it's a copy-paste fail. Fixed and pushed the new commit.

mikedanese commented:
@ixdy aaah, thanks.

madhusudancs commented:
Finally!

mikedanese commented:
mergebot might need a reboot

eparis commented:
@mikedanese blast, looks like someone bounced the bot, but kubectl logs -p isn't giving me the last set with info to debug what went wrong   :-(

eparis commented:
it seems to be working correctly now (and this wasn't merged because e2e failed, not because of a bot problem)

madhusudancs commented:
@eparis Did the e2e fail because of this PR? Is there a way I can look at the mergebot's e2e logs? I looked at http://submit-queue.k8s.io/ but I wasn't sure how to look at e2e logs.

mikedanese commented:
@madhusudancs it was not caused by this PR. When the critical builds at https://kubekins.dls.corp.google.com start to fail, the build queue stops. nothing to do with the PR. @eparis thanks for keeping an eye out!

madhusudancs commented:
@mikedanese Oh Ok, cool. Thanks for that pointer.

eparis commented:
Sorry about that, didn't mean to blame this PR, just trying to discuss what the bot is doing, since usually when the bot does things that seem weird, it was my fault.

madhusudancs commented:
@eparis no problem. I am still new to the project and learning new things everyday :) Given this PR's long history of e2e failures, the first thing that came to my mind was a failure due to this PR. But now I know what's going on. Learned something new today :)

mikedanese commented:
@k8s-bot ok to test
